File: frontend/src/models/dto/eventactions/CallUrlEventActionDto.ts
declare module 'CallUrlEventActionDto/Request' {
  import EventActionDtoRequest from 'EventActionDto/Request'

  export interface Create extends EventActionDtoRequest.Create {
    requestMethod: string;
    url: string;
    type: 'callUrl'
  }
}

declare module 'CallUrlEventActionDto/Response' {
  import EventActionDtoResponse from 'EventActionDto/Response'

  export interface Detailed extends EventActionDtoResponse.Detailed {
    requestMethod: string;
    url: string;
    type: 'callUrl'
  }
}

File: backend/src/main/java/net/alex9849/cocktailpi/payload/dto/eventaction/CallUrlEventActionDto.java
package net.alex9849.cocktailpi.payload.dto.eventaction;

import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.*;
import net.alex9849.cocktailpi.model.eventaction.CallUrlEventAction;
import org.springframework.web.bind.annotation.RequestMethod;

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class CallUrlEventActionDto {
    private interface RequestCommand { @NotNull RequestMethod getRequestMethod(); }
    private interface Url { @NotNull @Size(max = 255) String getUrl(); }

    @NoArgsConstructor(access = AccessLevel.PRIVATE)
    public static class Request {
        @Getter @Setter @EqualsAndHashCode(callSuper = true)
        public static class Create extends EventActionDto.Request.Create implements RequestCommand, Url {
            RequestMethod requestMethod;
            String url;
        }
    }

    @NoArgsConstructor(access = AccessLevel.PRIVATE)
    public static class Response {
        @Getter @Setter @EqualsAndHashCode(callSuper = true)
        public static class Detailed extends EventActionDto.Response.Detailed implements RequestCommand, Url {
            RequestMethod requestMethod;
            String url;

            protected Detailed(CallUrlEventAction eventAction) {
                super(eventAction);
            }

            public String getType() {
                return "callUrl";
            }
        }
    }
}
